File: Interview Problems/Parking Lot/Java Code/repository/PricingRuleRepository.java
package repository;

import domain.PricingRule;
import domain.Vehicle;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class PricingRuleRepository {

    // Keyed by VehicleType for direct O(1) lookup — one rule per vehicle type
    private final Map<Vehicle.VehicleType, PricingRule> store = new HashMap<>();

    public void save(PricingRule rule) {
        store.put(rule.getVehicleType(), rule);
        System.out.println("[REPO] PricingRule saved for: " + rule.getVehicleType());
    }

    public Optional<PricingRule> findByVehicleType(Vehicle.VehicleType vehicleType) {
        return Optional.ofNullable(store.get(vehicleType));
    }

    public void update(PricingRule rule) {
        if (!store.containsKey(rule.getVehicleType())) {
            throw new IllegalArgumentException("No pricing rule for vehicle type: " + rule.getVehicleType());
        }
        store.put(rule.getVehicleType(), rule);
        System.out.println("[REPO] PricingRule updated for: " + rule.getVehicleType());
    }
}

File: Interview Problems/Parking Lot/Python Code/repository/pricing_rule_repository.py
from typing import Dict, Optional
from domain.vehicle import VehicleType
from domain.pricing_rule import PricingRule

class PricingRuleRepository:
    def __init__(self):
        self._store: Dict[VehicleType, PricingRule] = {}

    def save(self, rule: PricingRule):
        self._store[rule.vehicle_type] = rule
        print(f"[REPO] PricingRule saved for: {rule.vehicle_type.name}")

    def find_by_vehicle_type(self, vehicle_type: VehicleType) -> Optional[PricingRule]:
        return self._store.get(vehicle_type)

    def update(self, rule: PricingRule):
        if rule.vehicle_type not in self._store:
            raise ValueError(f"No pricing rule for vehicle type: {rule.vehicle_type.name}")
        self._store[rule.vehicle_type] = rule
        print(f"[REPO] PricingRule updated for: {rule.vehicle_type.name}")
